File: backend/app/rag/factory.py
"""Construct the embedding provider and retriever from configuration."""

from functools import lru_cache
from pathlib import Path

from app.config import get_settings
from app.rag.embeddings import EmbeddingProvider, OpenAICompatibleEmbeddings
from app.rag.retriever import Retriever
from app.rag.store import VectorStore

_DOCUMENTS_FILE = "documents.json"


@lru_cache
def get_embeddings() -> EmbeddingProvider:
    settings = get_settings()
    return OpenAICompatibleEmbeddings(
        api_key=settings.embedding_api_key,
        base_url=settings.embedding_base_url,
        model=settings.embedding_model,
    )


@lru_cache
def get_retriever() -> Retriever:
    settings = get_settings()
    index_dir = Path(settings.index_dir)
    if (index_dir / _DOCUMENTS_FILE).exists():
        store = VectorStore.load(index_dir)
    else:
        # No index built yet: an empty store makes /chat degrade gracefully
        # to plain LLM answers (with honest "no knowledge base" hedging).
        store = VectorStore(dim=1, model=settings.embedding_model)
    return Retriever(
        store,
        get_embeddings(),
        top_k=settings.retrieval_top_k,
        min_score=settings.retrieval_min_score,
    )

File: lib/rag/factory.ts
/**
 * Construct the embedding provider and retriever from configuration.
 *
 * Port of `backend/app/rag/factory.py` (`@lru_cache` becomes a `globalThis`
 * cache so the instances survive Next.js dev hot reloads). The vector store
 * is shared between the chat route (retrieval) and the files add-to-kb route
 * (ingestion), so documents added at runtime are immediately retrievable.
 */

import fs from "node:fs";
import path from "node:path";

import { getSettings } from "@/lib/config";
import { OpenAICompatibleEmbeddings, type EmbeddingProvider } from "@/lib/rag/embeddings";
import { Retriever } from "@/lib/rag/retriever";
import { VectorStore } from "@/lib/rag/store";

const DOCUMENTS_FILE = "documents.json";

const globalCache = globalThis as typeof globalThis & {
  __renovationEmbeddings?: EmbeddingProvider;
  __renovationRetriever?: Retriever;
};

export function getEmbeddings(): EmbeddingProvider {
  if (globalCache.__renovationEmbeddings === undefined) {
    const settings = getSettings();
    globalCache.__renovationEmbeddings = new OpenAICompatibleEmbeddings({
      apiKey: settings.embeddingApiKey,
      baseUrl: settings.embeddingBaseUrl,
      model: settings.embeddingModel,
    });
  }
  return globalCache.__renovationEmbeddings;
}

function loadStore(): VectorStore {
  const settings = getSettings();
  if (fs.existsSync(path.join(settings.indexDir, DOCUMENTS_FILE))) {
    try {
      return VectorStore.load(settings.indexDir);
    } catch (error) {
      console.warn(`Ignoring unreadable vector index in ${settings.indexDir}:`, error);
    }
  }
  // No index built yet: an empty store makes chat degrade gracefully to
  // plain LLM answers (with honest "no knowledge base" hedging).
  return new VectorStore(1, settings.embeddingModel);
}

export function getRetriever(): Retriever {
  if (globalCache.__renovationRetriever === undefined) {
    const settings = getSettings();
    globalCache.__renovationRetriever = new Retriever(loadStore(), getEmbeddings(), {
      topK: settings.retrievalTopK,
      minScore: settings.retrievalMinScore,
    });
  }
  return globalCache.__renovationRetriever;
}

/**
 * Swap the vector store backing the shared retriever singleton.
 *
 * Used by add-to-kb when the placeholder empty store (created before any
 * index existed) must be replaced with one matching the real embedding
 * dimension; chat retrieval picks up the new store on the next request.
 */
export function replaceVectorStore(store: VectorStore): VectorStore {
  const settings = getSettings();
  globalCache.__renovationRetriever = new Retriever(store, getEmbeddings(), {
    topK: settings.retrievalTopK,
    minScore: settings.retrievalMinScore,
  });
  return store;
}

/** Drop the cached instances (tests, or after settings changes). */
export function resetRagSingletons(): void {
  delete globalCache.__renovationEmbeddings;
  delete globalCache.__renovationRetriever;
}
